presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -686,6 +718,44 @@ private List<TaskSource> getTaskSources(Iterator<SerializedPrestoSparkTaskSource
         return result.build();
     }
 
+    private List<TaskSource> getNativeExecutionShuffleSources(
+            Session session, TaskId taskId, PlanFragment fragment, Map<PlanNodeId, PrestoSparkShuffleReadInfo> shuffleReadInfos, List<TaskSource> taskSources)
+    {
+        if (!isNativeExecutionEnabled(session)) {
+            return taskSources;
+        }
+
+        // Could be COORDINATOR_ONLY fragment or hybrid execution fragment in future
+        if (!(fragment.getRoot() instanceof NativeExecutionNode)) {
+            return taskSources;
+        }
+
+        PlanNode root = fragment.getRoot();
+        List<TaskSource> nativeExecutionSources = taskSources.stream().filter(taskSource -> isNativeExecutionNode(taskSource.getPlanNodeId(), root)).collect(Collectors.toList());
+        checkArgument(nativeExecutionSources.size() == 1, "Only 1 taskSource is expected for NativeExecutionNode but got %s", nativeExecutionSources.size());
+
+        ImmutableSet.Builder<ScheduledSplit> result = ImmutableSet.builder();
+        AtomicLong nextSplitId = new AtomicLong();
+        shuffleReadInfos.forEach((planNodeId, info) ->
+                result.add(new ScheduledSplit(nextSplitId.getAndIncrement(), planNodeId, new Split(REMOTE_CONNECTOR_ID, new RemoteTransactionHandle(), new RemoteSplit(
+                        new Location(shuffleInfoSerializer.serializeReadInfo(info)),
```

Comment:
FYI some native engine code will be checked in soon. The location should be of some format. Below is the documentation for exchange source creation:

```
  /// url needs to follow below format:
  /// batch://<taskid>?shuffleInfo=<serialized-shuffle-info>
  static std::unique_ptr<velox::exec::ExchangeSource> createExchangeSource(
      const std::string& url,
      int32_t destination,
      std::shared_ptr<velox::exec::ExchangeQueue> queue,
      velox::memory::MemoryPool* FOLLY_NONNULL pool);
```